src/borrow_tracker/tree_borrows/perms.rs

```diff
@@ -271,14 +265,99 @@ pub mod diagnostics {
             match (self.0, self.1) {
-                (_, Active) => "an activation",
+                (_, Active) => "the first write to a 2-phase borrowed mutable reference",
                 (_, Frozen) => "a loss of write permissions",
                 (Frozen, Disabled) => "a loss of read permissions",
                 (_, Disabled) => "a loss of read and write permissions",
                 (old, new) =>
                     unreachable!("Transition from {old:?} to {new:?} should never be possible"),
             }
         }
+
+        /// Not all transitions are relevant to all errors, this function filters those out.
+        ///
+        /// For example we don't care about the activation of a pointer when the error is
+        /// because it is not readable. All we care about is when it lost read permissions.
+        /// Return value to be interpreted as
+        /// - `None`: not a relevant transition
+        /// - `Some(v)`: relevant transition, with the additional help messages `v`
+        ///
+        /// Irrelevant events:
+        /// - modifications of write permissions when the error is related to read permissions
+        ///   (on failed reads and protected `Frozen -> Disabled`, ignore `Reserved -> Active`,
+        ///   `Reserved -> Frozen`, and `Active -> Frozen`)
+        /// - all transitions for attempts to deallocate strongly protected tags
+        ///
+        /// # Panics
+        ///
+        /// This function assumes that its arguments apply to the same location
+        /// and that they were obtained during a normal execution. It will panic otherwise.
+        /// - `err` cannot be a `ProtectedTransition(_)` of a noop transition, as those
+        /// never trigger protectors;
+        /// - all transitions involved in `self` and `err` should be increasing
+        /// (Reserved < Active < Frozen < Disabled);
+        /// - between `self` and `err` the permission should also be increasing,
+        /// so all permissions inside `err` should be greater than `self.1`;
+        /// - `Active` and `Reserved` cannot cause an error due to insufficient permissions,
+        /// so `err` cannot be a `ChildAccessForbidden(_)` of either of them;
+        pub(in super::super) fn is_relevant(&self, err: TransitionError) -> Option<Vec<String>> {
+            // NOTE: `super::super` is the visibility of `TransitionError`
+            assert!(Self::is_possible(self.0, self.1));
+            if self.is_noop() {
+                return None;
+            }
+            match err {
+                TransitionError::ChildAccessForbidden(insufficient) => {
+                    // Show where the permission was gained then lost,
+                    // but ignore unrelated permissions.
+                    // This eliminates transitions like `Active -> Frozen`
+                    // when the error is a failed `Read`.
+                    match (self.0, self.1, insufficient.0) {
```

Comment:
`self.0` is never actually matched here, is it?

Also I have no idea what all these permissions mean.^^ I guess self.1 is the `to` permission of this transition. What's `insufficient.0`?

(That is the problem with tuple structs: the fields have no names that could explain their meaning.)